### PR TITLE
docs(contributing): add contribution and new-container guide 

### PR DESCRIPTION
**Pull request by totallyzen**
# change

Document how to contribute, with initial focus on making local development smooth.

# Tasks
- [x] Finish the `new-container` guide
- [x] Remove any old docs referring to 
- [x] Update `README.md` to point at the contribution guide
- [x] Update `README.md` to add badges (supported python versions, etc) and to give kudos to current and past maintainers and contributors

### CODE REVIEW

**Comment by totallyzen** [comment]
@jankatins @bearrito @max-pfeiffer your comments are very welcome - It's still in draft, but it's a start!

This is what I had time for today, but will continue on the weekday evenings!

**Comment by totallyzen** [comment]
@santi also, this will be useful for you!

File: .github/CONTRIBUTING.md
Comment (bearrito) [inline .github/CONTRIBUTING.md:30]: I'm using the devcontainers workflow. I could add in a section on using that if you wanted.

File: .github/ISSUE_TEMPLATE/new-container.md
Comment (bearrito) [inline .github/ISSUE_TEMPLATE/new-container.md:16]: nit: Why not just use a generic container **for** this?

File: .github/ISSUE_TEMPLATE/new-container.md
Comment (bearrito) [inline .github/ISSUE_TEMPLATE/new-container.md:18]: L -> lowercase

**Comment by alexanderankin** [comment]
looks good to me, i find this stuff very valuable but right now i dont really know what to add as far as suggestions or discussion. my head is mostly in the outstanding PRs and releases and such.

also sorry for the double rebase. i did a rebase locally onto main. then i found a button on github, but the button on github gave me credit for all your work, so then i just pushed my local rebase.

**Review by alexanderankin — APPROVED** [review]

**Comment by alexanderankin** [comment]
it might also make sense to capture what it takes to implement a new container, not just request one

File: .github/ISSUE_TEMPLATE/new-container.md
Comment (santi) [inline .github/ISSUE_TEMPLATE/new-container.md:14]: -fit*

File: .github/CONTRIBUTING.md
Comment (totallyzen) [reply .github/CONTRIBUTING.md:30]: hey @bearrito apologies for the delays, I'm finding myself with some time today.
I'd love your contribution on the devcontainers workflow, but lemme finish this PR (ETA today, finally found the time to work on this) and then yo can raise your own PR for that addition to the docs :pray:

**Comment by alexanderankin** [comment]
quick summary of how to test buildthedocs changes:

1. make a new project on buildthedocs
1. "import" your fork of this repository
1. push changes and test there
    1. significant files: `/.readthedocs.yml`, `/Makefile` (the `make docs` target, specifically), `/conf.py` (the sphinx stuff), `/docs/_build` (preview the result locally)
1. then open pr here

File: Makefile
Comment (max-pfeiffer) [inline Makefile:30]: Do we already have a plan what to do with our test coverage reports?

We could export it to [codecov](https://about.codecov.io/) and add a badge to our README.md.

**Comment by max-pfeiffer** [comment]
> it might also make sense to capture what it takes to implement a new container, not just request one

Yes, I think so too. We could briefly elaborate about the best practices that we want to see in contributions. So modules are a bit more consistent. 

Maybe also what to include in README.rst in the modules directory. To me it's not clear what purpose that serves currently.

**Comment by max-pfeiffer** [comment]
I would say also our [.pre-commit-config.yaml](https://github.com/testcontainers/testcontainers-python/blob/main/.pre-commit-config.yaml) also needs some attention: there we have black for code formatting and ruff as linter defined.
What about just using ruff? ruff can format code as well and is pretty fast in doing that. Then we can drop that black dependency.

I like badges :smiley:. So would could add the following badges:
- Poetry
- ruff
- supported Python versions
- test coverage Codecov, see my comment on the make file for running tests with pytest-cov

@totallyzen Can I add commits to this PR? Then I could quickly contribute my suggestions.

File: Makefile
Comment (totallyzen) [reply Makefile:30]: :wave: yep, I'd like that, but it needs @kiview's involvement as you need to install the GH app - I am missing some privileges that we'll be needing to install different apps.

We'll talk about this on slack I think.

**Comment by totallyzen** [comment]
> Can I add commits to this PR? Then I could quickly contribute my suggestions.

Yes, please do!
Including the `black` -> `ruff` transition _as long as it doesn't reformat the entire repo_ :slightly_smiling_face: 
If we are to improve/fix formatting, I want that to be in another repo.

@bearrito actually, I think given my lack of time lately, you could contribute to the `devcontainers` to this PR as well, just ping the request on this PR as well so I can track :pray:

**Comment by jankatins** [comment]
> Including the black -> ruff transition as long as it doesn't reformat the entire repo 🙂

Ruff is compatible with black apart from a very few and rare differences. I've several repos here where black and ruff format are interchangeable :-)

File: .github/CONTRIBUTING.md
Comment (jankatins) [inline .github/CONTRIBUTING.md:15]:
```suggestion
1. Finish reading this document.
```

File: .github/CONTRIBUTING.md
Comment (jankatins) [inline .github/CONTRIBUTING.md:33]: If someone is not wanting to use pre-commit, there could also be a 'lint' makefile target which runs it on all files

```Makefile
pre-commit-all:
    poetry run pre-commit run --all-files
```
(replace 4 spaces with tab...)

File: Makefile
Comment (jankatins) [inline Makefile:25]: As someone not familiar with what "dist" means, I would prefer a more speaking name.

File: Makefile
Comment (jankatins) [inline Makefile:73]: :heart:

**Comment by bearrito** [comment]
@totallyzen I can do that. It will either be today or tomorrow. Likely tomorrow.

**Comment by bearrito** [comment]
@totallyzen Can I get this approved if it looks okay- https://github.com/testcontainers/testcontainers-python/pull/506 Namely I want to be able to reference that work in the docs for devcontainers

**Comment by MattOates** [comment]
As someone who just tried to raise a PR with a new container, and just checked out these new docs. Can I point out its not obvious at all on these points:

> .github/workflows/main.yml to run tests, build, and publish your package when pushed to the main branch.
No such file exists.

> Rebase your development branch on main (or merge main into your development branch).
Main just changed whilst I've had my PR up, is the preference I rebase and force push over the top of my already published PR constantly?

>  Add a line -e file:[feature name] to requirements.in
Where is `requirements.in` ??? If you add directly to pyproject.toml under a module is this sufficient and correct?

**Comment by totallyzen** [comment]
Hey @MattOates,

Thanks for the pointers, I'm sorry for your bad experience, but please note that this PR is a draft, I did not have time to finish it. A lot of people left comments on it, all duly noted, you can in fact raise a PR against this branch and make suggestions. :+1: 
You can also keep your calm about it, the current maintainers are volunteers, with sometimes very little time to spare on this. :+1:

**Comment by totallyzen** [comment]
@bearrito done, I merged #506 - feel free to reference it in your next PR :pray: 

Apologies for the continued delays, my work weeks have been intense/depressing/exhausting. :bow:

File: Makefile
Comment (totallyzen) [reply Makefile]: dunzo, now we have just `build` and it does generate an artifact, removed it from this list.

File: .github/CONTRIBUTING.md
Comment (totallyzen) [reply .github/CONTRIBUTING.md:33]: done, but a few lines below your comment

File: .github/CONTRIBUTING.md
Comment (santi) [inline .github/CONTRIBUTING.md:69]: Should this one be a `[!WARNING]` or `[!CAUTION]`?

Comment (santi) [reply .github/CONTRIBUTING.md:69]: Woops, onyl checked changes from a single commit. Disregard this comment.

File: index.rst
Comment (santi) [inline index.rst:129]: This is no longer the case, should be changed to a text specifying `extras`

**Comment by santi** [comment]
Hey @MattOates sorry for your rough start here, but we really appreciate you contributions 😊 Hopefully the latest additions to the draft should make clear the things you are wondering about. 

We have just revamped the whole structure and development setup in this repo, so no wonder the docs are very outdated. I hope you stay and find the new setup easier to work with 😀

**Review by santi — APPROVED** [review]
A few last comments from my end, but I think the current draft is in a good shape and ready to be released. Further contributions are of course welcome and can be added in future PRs.

Great job here @totallyzen!

File: .github/CONTRIBUTING.md
Comment (alexanderankin) [reply .github/CONTRIBUTING.md:69]: i am personally okay with this one

File: .github/PULL_REQUEST_TEMPLATE/new_container.md
Comment (alexanderankin) [inline .github/PULL_REQUEST_TEMPLATE/new_container.md:40]:
```suggestion
- If the intended usage of the new Testcontainer is with a client library, please add the client library as a dev dependency, so that tests can mirror actual usage of the new container.
```

not sure if this is the best wording but maybe we encourage people to add dev dependencies instead of writing tests that never run

File: index.rst
Comment (alexanderankin) [inline index.rst:129]:
```suggestion
e.g., :code:`testcontainers[mysql]` and :code:`testcontainers[postgres]` for MySQL and PostgreSQL database containers, respectively.
```

**Comment by max-pfeiffer** [comment]
> Yes, please do! Including the `black` -> `ruff` transition _as long as it doesn't reformat the entire repo_ 🙂 If we are to improve/fix formatting, I want that to be in another PR.

@totallyzen I just created a PR and removed that black dependency: https://github.com/testcontainers/testcontainers-python/pull/529

Just a couple of files became re-formatted. So this we have rather little code changes.

File: .github/PULL_REQUEST_TEMPLATE/new_container.md
Comment (jankatins) [reply .github/PULL_REQUEST_TEMPLATE/new_container.md:40]: `... please add _such a_ client library as a dev dependency _and a test which mirrors actual usage of the container._` (e.g. PG has multiple which consume the url)

**Comment by totallyzen** [comment]
I'll merge this, I heard no negative comments on the existing additions and some improvements rely on `main` -> let's consider this PR to be "done"

**Comment by alexanderankin** [comment]
lgtm!

File: Makefile
Comment (alexanderankin) [reply Makefile:23]: spit into #544

Comment (alexanderankin) [reply Makefile:23]: *split

**Comment by MattOates** [comment]
> Hey @MattOates sorry for your rough start here, but we really appreciate you contributions 😊 Hopefully the latest additions to the draft should make clear the things you are wondering about.

Apologies if that came across as brusk, if I did know anything about the desired dev setup I'd absolutely help dig in. I just cargo culted my way to a PR and just wanted to point out the draft still had some rabbit holes for someone coming cold. Is there a shaping doc anywhere of where the project is headed for dev experience?

**Comment by alexanderankin** [comment]
if someone wants to help out - let me know if you see any glaring differences between these two pages:

https://github.com/testcontainers/testcontainers-python/compare/main..docs/improve-contribution-guide.tmp

- this is the rebased version of this PR - it needs the committer reset to the author and pushed

https://github.com/testcontainers/testcontainers-python/compare/main...docs/improve-contribution-guide

- this is a simplified view of the files tab on this page

File: .github/PULL_REQUEST_TEMPLATE/new_container.md
Comment (jankatins) [inline .github/PULL_REQUEST_TEMPLATE/new_container.md:37]: type: trying

**Pull request merged**
